src/zinoargus/__init__.py

```diff
@@ -323,6 +327,42 @@ def update_case_acknowledged(incident: Incident, case: ritz.Case, desired_state:
     _zino.set_state(case_id, desired_state)
 
 
+def update_case_ticket(incident: Incident, case_id: int):
+    """Updates a Zino case history with a new ticket URL from Argus, if it isn't there
+    already.
+    """
+    if not incident.ticket_url:
+        return
+    if is_string_in_case_history(case_id, incident.ticket_url):
+        return
+    user = find_who_added_incident_ticket(incident, incident.ticket_url)
+    _logger.info(
+        "New ticket URL %s found on Argus incident %s, adding to Zino case %s",
+        incident.ticket_url,
+        incident.pk,
+        case_id,
+    )
+    message = f"Ticket {incident.ticket_url} added by Argus user {user}"
+    _zino.add_history(case_id, message)
+
+
+def is_string_in_case_history(case_id: int, string: str) -> bool:
+    """Returns True if a substring is present in the case history"""
+    history = _zino.get_history(case_id)
+    return any(any(string in line for line in entry["log"]) for entry in history)
+
+
+def find_who_added_incident_ticket(incident: Incident, ticket_url: str) -> str:
+    """Traverses the incident events to find the user who added the ticket URL."""
+    for event in _argus.get_incident_events(incident):
+        if (
+            event.type == INCIDENT_ATTRIBUTE_CHANGE_TYPE
+            and ticket_url in event.description
```

Comment:
There is an edge case here that needs to be avoided:

User A changed ticket url from str_a to str_b
User B changed ticket url from str_b to str_a

When you search for it there will be two events with the descriptions
A: Change: "ticket_url": str_a → str_b
B: Change: "ticket_url": str_b → str_a

The order of the events is from last to first, so we will get the correct event and the correct user, but I'm not sure if I like it not being explicit

We could for example use `event.description.endswith(ticket_url)`

Reply:
I don't really want to put a lot of stock into the organization of the event description, since it is free-form text and could change in Argus.

That being said, this *does* assume the event list received from Argus is sorted descending by time (which it was in my tests).  This should mean the first event we see that contains the current ticket URL value in its description represents the last change to that value.  My assumption about the ordering could be incorrect (but the list could easily be sorted by time before processing).

Reply:
No, your assumption is correct, it is sorted descending by time, so the behavior is as expected

I would also be happy with a comment stating that, just in case that ever were to change and then we would totally clueless about a bug appearing